packages/rum-recorder/src/domain/rrweb-snapshot/snapshot.ts

```diff
@@ -0,0 +1,589 @@
+/* eslint-disable no-underscore-dangle */
+/* eslint-disable @typescript-eslint/prefer-regexp-exec */
+import {
+  SerializedNode,
+  SerializedNodeWithId,
+  NodeType,
+  Attributes,
+  INode,
+  IdNodeMap,
+  MaskInputOptions,
+  SlimDOMOptions,
+} from './types'
+
+const tagNameRegex = RegExp('[^a-z1-6-_]')
+
+export const IGNORED_NODE = -2
+
+let nextId = 1
+function genId(): number {
+  return nextId++
+}
+
+export function cleanupSnapshot() {
+  // allow a new recording to start numbering nodes from scratch
+  nextId = 1
+}
+
+function getValidTagName(tagName: string): string {
+  const processedTagName = tagName.toLowerCase().trim()
+
+  if (tagNameRegex.test(processedTagName)) {
+    // if the tag name is odd and we cannot extract
+    // anything from the string, then we return a
+    // generic div
+    return 'div'
+  }
+
+  return processedTagName
+}
+
+function getCssRulesString(s: CSSStyleSheet): string | null {
+  try {
+    const rules = s.rules || s.cssRules
+    return rules ? Array.from(rules).map(getCssRuleString).join('') : null
+  } catch (error) {
+    return null
+  }
+}
+
+function getCssRuleString(rule: CSSRule): string {
+  return isCSSImportRule(rule) ? getCssRulesString(rule.styleSheet) || '' : rule.cssText
+}
+
+function isCSSImportRule(rule: CSSRule): rule is CSSImportRule {
+  return 'styleSheet' in rule
+}
+
+function extractOrigin(url: string): string {
+  let origin
+  if (url.indexOf('//') > -1) {
+    origin = url.split('/').slice(0, 3).join('/')
+  } else {
+    origin = url.split('/')[0]
+  }
+  origin = origin.split('?')[0]
+  return origin
+}
+
+const URL_IN_CSS_REF = /url\((?:(')([^']*)'|(")([^"]*)"|([^)]*))\)/gm
+const RELATIVE_PATH = /^(?!www\.|(?:http|ftp)s?:\/\/|[A-Za-z]:\\|\/\/).*/
+const DATA_URI = /^(data:)([^,]*),(.*)/i
+export function absoluteToStylesheet(cssText: string | null, href: string): string {
```

Comment:
What about also importing unit tests?
cf https://github.com/rrweb-io/rrweb-snapshot/blob/master/test/snapshot.test.ts

Reply:
fair enough, although this adds a dependency on jsdom and chai, are we comfortable with that? I know @BenoitZugmeyer wanted to look at importing rrweb tests in another PR

Reply:
Yes, I'm working on it. We should not import chai or jsdom though:
* jsdom is used when running tests in Nodejs (through Jest). Since we use Karma, we can just use the real DOM
* chai is an assertion library, karma-jasmine provides one out of the box

Reply:
I think importing snapshot unit tests would be really straightforward as there is no snapshots or complex asynchronous cases. It could be worth it to import and adjust them (to use karma/jasmine) here.